Update mdbook to 0.4.37

ehuss opened the pull request:
This updates mdbook to 0.4.37.
Changelog: https://github.com/rust-lang/mdBook/blob/master/CHANGELOG.md#mdbook-0437

The primary change is the update to pulldown-cmark which has a large number of markdown parsing changes. There shouldn't be any significant changes to the rendering of any of the books (I have posted some PRs to fix some minor issues to the ones that were affected).

ehuss committed:
Update mdbook to 0.4.37

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7809911578/job/21302490748) [(plain)](https://github.com/rust-lang/rust/commit/51f20b4f6ba7ba116dcd9bfd7ff328dc163d268e/checks/21302490748/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a2869ca9-01f0-4364-b975-8b2e9efa83d4
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=update-mdbook
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_a2869ca9-01f0-4364-b975-8b2e9efa83d4
GITHUB_REF=refs/pull/120729/merge
GITHUB_REF_NAME=120729/merge
GITHUB_REF_PROTECTED=false
---
#12 writing image sha256:b3fac7a4f9a429d45c234ba6f95d8134ab21d4eaa0460310d6143057a5a607c8 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Feb  7 04:27:17 UTC 2024
  network time: Wed, 07 Feb 2024 04:27:17 GMT
  network time: Wed, 07 Feb 2024 04:27:17 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished release [optimized] target(s) in 1.35s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --no-default-features --mini-tests --std-tests`
Using system GCC
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.10s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
rustdoc/write-documentation/the-doc-attribute.html:227: broken link - `rustdoc/write-documentation/{}`
rustdoc/print.html:446: broken link - `rustdoc/foo.css`
rustdoc/print.html:1301: broken link - `rustdoc/{}`
rustdoc/command-line-arguments.html:334: broken link - `rustdoc/foo.css`
number of HTML files scanned: 34989
number of HTML redirects found: 10559
number of links checked: 3079833
number of links ignored due to external: 67531
```

</details>

ehuss commented:
I'm still working on fixing the test issue.

@rustbot author

ehuss committed:
Add test for broken_intra_doc_link.

ehuss committed:
Switch linkchecker to use html5ever for html parsing.

The existing regex-based HTML parsing was just too primitive to
correctly handle HTML content. Some books have legitimate `href="…"`
text which should not be validated because it is part of the text, not
actual HTML.

ehuss commented:
@rustbot ready

OK, this is ready now. I added some commits to update the linkchecker to use an actual HTML parser instead of a regex. This was necessary because the new pulldown_cmark changed some escaping behaviors, and the regex couldn't handle random `href="` within normal text. There is a performance penalty, going from about 6 seconds to 18 seconds, but should still be tolerable. I'm not sure if html5ever is really the best choice, but it works well enough I think.

Comparison of the behavior:

Before:
checked links in: 6.7s
number of HTML files scanned: 35034
number of HTML redirects found: 10570
number of links checked: 3009436
number of links ignored due to external: 143040
number of links ignored due to exceptions: 10
number of intra doc links ignored: 7
errors found: 0

After:
checked links in: 18.8s
number of HTML files scanned: 35034
number of HTML redirects found: 10570
number of links checked: 3009439
number of links ignored due to external: 143040
number of links ignored due to exceptions: 10
number of intra doc links ignored: 7
errors found: 0

Mark-Simulacrum commented:
Performance hit is unfortunate but seems indeed acceptable.

@bors r+

bors commented:
:pushpin: Commit 776590b14edbc3bc840d419035ca8deaadaeeda1 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 776590b14edbc3bc840d419035ca8deaadaeeda1 -->
<!-- homu: {"type":"Approved","sha":"776590b14edbc3bc840d419035ca8deaadaeeda1","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
